[DRAFT] feat: puya debugging support; improved source map management (#37)
algorandfoundation/algokit-avm-vscode-debugger: README.md

In order to use the AVM Debugger extension, you need:

1. Simulate Traces. A `*.trace.avm.json` file that maps the compiled TEAL source maps to the original source code. See an example [here](./examples/multiRootWorkspace/slot-machine/debug_traces/simulate-response.trace.avm.json).

> 2. TEAL Source Maps. A `sources.avm.json` file that contains the traces obtained from algod's [`simulate` endpoint](https://developer.algorand.org/docs/get-details/dapps/smart-contracts/debugging/?from_query=simulate#simulate). This serves as an entry point for the debugger. See an example [here](./examples/multiRootWorkspace/slot-machine/.algokit/sources/sources.avm.json).

Reviewer: I'm wondering if we should even document this, given that we dynamically build it rather than requiring it to launch.

Reviewer 2: I think just the descriptions should we reworked. The first part of each point is correct, but the description for each is incorrect

e.g.

```
1. Simulate Traces. A `*.trace.avm.json`  file that contains the traces obtained from algod's [`simulate` endpoint](https://developer.algorand.org/docs/get-details/dapps/smart-contracts/debugging/?from_query=simulate#simulate). This serves as an entry point for the debugger. See an example [here](./examples/multiRootWorkspace/slot-machine/debug_traces/simulate-response.trace.avm.json).
2. Source Maps. Source maps for the programs executed within the trace. These can be either a 
    * a TEAL source map, obtained from algod when compiling a TEAL program
    * a Puya source map, produced when compiling a Smart Contract with the Puya compiler
```

Reviewer: Yeah agreed

#### Python

```py
# Place this code in a project entry point (e.g. main.py)
from algokit_utils.config import config
config.configure(debug=True, trace_all=True)
```

#### TypeScript

If your project instance does not contain `algorandfoundation/algokit-utils-debug` in package.json, follow the steps below, otherwise feel free to skip remaining steps.

1. Run `npm install @algorandfoundation/algokit-utils-debug` to install the debug extension.
2. Add the following code to your project entry point (e.g. index.ts).

```ts
// Place this code in a project entry point (e.g. index.ts)
import { Config } from '@algorandfoundation/algokit-utils'
import { registerDebugEventHandlers() } from '@algorandfoundation/algokit-utils-debug'

Config.configure({
  debug: true,
  traceAll: true,
})
registerDebugEventHandlers() // IMPORTANT: must be called before any transactions are submitted.
```

For more details on supplementary debug package for algokit-utils-ts, refer to package [repository](https://github.com/algorandfoundation/algokit-utils-ts-debug).

> > NOTE: Storing debug traces is not possible in browser environments, your contract project needs access to filesystem via `node`. If you wish to extract simulate traces manually from frontend app that relies on `algokit-utils-ts`, refer to dedicated section in [algokit-utils-ts docs](https://github.com/algorandfoundation/algokit-utils-ts/blob/main/docs/capabilities/debugging.md).

Reviewer: ```suggestion
> NOTE: Storing debug traces is not possible in browser environments, your contract project needs access to filesystem via `node`. If you wish to extract simulate traces manually from an app running in a browser that uses `algokit-utils-ts`, refer to [algokit-utils-ts docs](https://github.com/algorandfoundation/algokit-utils-ts/blob/main/docs/capabilities/debugging.md#debugging-in-browser-environment).
```
